Sources/News.php

```diff
@@ -18,23 +18,76 @@
 
 /**
  * Outputs xml data representing recent information or a profile.
- * Can be passed 4 subactions which decide what is output:
+ *
+ * Can be passed subactions which decide what is output:
  *  'recent' for recent posts,
  *  'news' for news topics,
  *  'members' for recently registered members,
  *  'profile' for a member's profile.
- * To display a member's profile, a user id has to be given. (;u=1)
- * Outputs an rss feed instead of a proprietary one if the 'type' $_GET
- * parameter is 'rss' or 'rss2'.
+ *  'posts' for a member's posts.
+ *  'pms' for a member's personal messages.
+ *
+ * When displaying a member's profile or posts, the u parameter identifies which member. Defaults
+ * to the current user's id.
+ * To display a member's personal messages, the u parameter must match the id of the current user.
+ *
+ * Outputs can be in RSS 0.92, RSS 2, Atom, RDF, or our own custom XML format. Default is RSS 2.
+ *
  * Accessed via ?action=.xml.
- * Does not use any templates, sub templates, or template layers.
  *
- * @uses Stats language file.
+ * Does not use any templates, sub templates, or template layers...
+ * ...except when requesting all the user's own posts or PMs. Then we show a template indicating
+ * our progress compiling the info. This template will auto-refresh until the all the info is
+ * compiled, at which point we emit the full XML feed as a downloadable file.
+ *
+ * @uses Stats language file, and in special cases the Admin template and language file.
  */
 function ShowXmlFeed()
 {
 	global $board, $board_info, $context, $scripturl, $boardurl, $txt, $modSettings, $user_info;
-	global $query_this_board, $smcFunc, $forum_version, $settings;
+	global $query_this_board, $smcFunc, $forum_version, $settings, $cachedir;
+
+	// List all the different types of data they can pull.
+	$subActions = array(
+		'recent' => array('getXmlRecent', 'recent-post'),
+		'news' => array('getXmlNews', 'article'),
+		'members' => array('getXmlMembers', 'member'),
+		'profile' => array('getXmlProfile', null),
+		'posts' => array('getXmlPosts', 'member-post'),
+		'pms' => array('getXmlPMs', 'personal-message'),
+	);
+
+	// Easy adding of sub actions
+	call_integration_hook('integrate_xmlfeeds', array(&$subActions));
+
+	if (empty($_GET['sa']) || !isset($subActions[$_GET['sa']]))
+		$_GET['sa'] = 'recent';
+
+	// Users can always export their own profile data
+	if (in_array($_GET['sa'], array('profile', 'posts', 'pms')) && !$user_info['is_guest'] && (empty($_GET['u']) || (int) $_GET['u'] == $user_info['id']))
```

Comment:
sutch width get hard to read,
we should consider to define a max length(something which flow in my mind when i see this).
But nothing which needs to change here.